lib/Dialect/TTNN/Transforms/TTNNWorkarounds.cpp

```diff
@@ -0,0 +1,418 @@
+
+  LogicalResult matchAndRewrite(wa::TTNNWorkaroundInterface op,
+                                PatternRewriter &rewriter) const final {
+
+    // To layout op is a special case, we don't want to rewrite it
+    if (mlir::isa<ttnn::ToLayoutOp>(op.getOperation())) {
+      return failure();
+    }
+
+    // Get the operands workarounds for the current operation
+    wa::TTNNOperandsWorkarounds operandsWorkarounds =
+        op.getOperandsWorkarounds();
+
+    // Apply input workarounds only for tensor operands
+    bool modifiedOperands = false;
+    int input_operand_index = 0;
+    for (size_t i = 0;
+         i < operandsWorkarounds.getInputOperandWorkarounds().size(); i++) {
+      wa::TTNNOperandWorkarounds inputWorkaround =
+          operandsWorkarounds.getInputOperandWorkarounds()[i];
+
+      // No input operand workarounds to apply, hence continue
+      if (!inputWorkaround.hasAnyWorkaround()) {
+        input_operand_index++;
+        continue;
+      }
+
+      // Get the next tensor opearand
+      while (!mlir::isa<RankedTensorType>(
+          op->getOperand(input_operand_index).getType())) {
+        input_operand_index++;
+      }
+
+      OpOperand &inputOperand = op->getOpOperand(input_operand_index++);
+      // Apply all workaround changes to the input operand
+      modifiedOperands |=
+          workaroundInputOperand(inputOperand, inputWorkaround, rewriter, op);
+    }
+
+    // Apply output workarounds only for tensor operands
+    int output_operand_index = 0;
+    for (size_t i = 0;
+         i < operandsWorkarounds.getOutputOperandWorkarounds().size(); i++) {
+      wa::TTNNOperandWorkarounds outputWorkaround =
+          operandsWorkarounds.getOutputOperandWorkarounds()[i];
+
+      // No output operand workarounds to apply, hence continue
+      if (!outputWorkaround.hasAnyWorkaround()) {
+        output_operand_index++;
+        continue;
+      }
+
+      // Get the next tensor result
+      while (!mlir::isa<RankedTensorType>(op->getResult(i).getType())) {
+        output_operand_index++;
+      }
+
+      OpResult outputResult = op->getResult(output_operand_index++);
+      // Apply all workaround changes to the output operand
+      modifiedOperands |=
+          workaroundOutputOperand(outputResult, outputWorkaround, rewriter, op);
+    }
+
+    return modifiedOperands ? success() : failure();
+  }
+};
+
+// Pass to apply workarounds to the operands of TTNN operations.
+class TTNNWorkarounds : public impl::TTNNWorkaroundsBase<TTNNWorkarounds> {
+public:
+  using impl::TTNNWorkaroundsBase<TTNNWorkarounds>::TTNNWorkaroundsBase;
+
+  void runOnOperation() final {
+    {
+      // Placeholder for workaround decomposition patterns
+    }
+    {
+      RewritePatternSet patterns(&getContext());
+      patterns.add<TTNNOperandsWorkaroundsRewriter>(&getContext());
+
+      FrozenRewritePatternSet patternSet(std::move(patterns));
+      GreedyRewriteConfig config = GreedyRewriteConfig();
```

Comment:
I took a look at the GreedyRewriter a bit, it seems that by default it is used with max of 10 iterations. However there is a config that specifies no limit in iterations, hence it will apply changes until the driver gets into the complete state:
config.maxIterations = GreedyRewriteConfig::kNoLimit;

I am more for using the driver as it gives a lot of things out of the box for free (transactional changes, working queue, ...).

Regrading what the Nick said, I rewrote the code that uses the rewriter to propagate changes. In the driver, the IR must be modified using the rewriter API and not directly.

Reply:
I thought I remembered reading somewhere that `matchAndRewrite` was only allowed to modify the matched op itself by either adding new ops around it, erasing it, or modifying it in place, but it was illegal to traverse somewhere else and update some other op.  I'm having trouble digging up where I read this though.

@mrakitaTT, I think you hit something similar using pattern rewriter during dialect conversion.  Do you remember any constraints related to this?

Reply:
I belive what you are refering to are operation passes documentation:
https://mlir.llvm.org/docs/PassManagement/#operation-pass

There are limitations that needs to be followed when executing operation passes becase they run in parallel. But the thing here is that we are defining this pass (and most of other our passes) on a module op, hence this limitations don't apply as we have only one module op.

Regarding pattern rewriter, the only limitation that I found is that all IR modifications needs to be done under rewriter object itself in order to keep changes sync in both rewriter and IR.

Reply:
Just to update that I rewrote the rewriter to not propagate the changes to neighbour ops. The code now propagate DPS changes for the current op.